capa/ghidra/capa_explorer.py

```diff
@@ -0,0 +1,261 @@
+            return hex(node_dict)
+
+        if isinstance(node_dict, str):
+            # expect the "description" key's string value
+            return node_dict
+
+        if "description" in node_dict:
+            return node_dict.get("description")
+        else:
+            # if no "description" key, the "type" key's value is the
+            # expected key.
+            # ex. {'type':'api', 'api':'RegisterServiceCtrlHandler'}
+            return self.recurse_node(node_dict.get(node_dict.get("type")))
+
+    def create_namespace(self):
+        """create new ghidra namespace for each capa namespace"""
+
+        # handle rules w/o namespace -> capa lib rule
+        if self.namespace == "capa":
+            lib_str = Namespace.DELIMITER + "lib" + Namespace.DELIMITER
+            self.namespace = self.namespace + lib_str + self.capability.replace(" ", "-")
+        cmd = CreateNamespacesCmd(self.namespace, SourceType.USER_DEFINED)
+        cmd.applyTo(currentProgram())  # type: ignore [name-defined] # noqa: F821
+
+        return cmd.getNamespace()
+
+    def tag_functions(self):
+        """create function tags for capabilities"""
+
+        # self.locations[0] will always be the largest
+        # scoped offset yielded i.e. closest to an entrypoint
+        addr = toAddr(hex(self.locations[0]))  # type: ignore [name-defined] # noqa: F821
+        func = getFunctionContaining(addr)  # type: ignore [name-defined] # noqa: F821
+
+        # bookmark Mitre ATT&CK tactics & MBC @ function scope
+        if func:
+            func.addTag(self.capability)
+            for item in self.attack:
+                attack_txt = item.get("tactic") + Namespace.DELIMITER + item.get("id")
+                add_bookmark(addr, attack_txt, "CapaExplorer::Mitre ATT&CK")
+
+            for item in self.mbc:
+                mbc_txt = item.get("objective") + Namespace.DELIMITER + item.get("id")
+                add_bookmark(addr, mbc_txt, "CapaExplorer::MBC")
+
+    def bookmark_locations(self):
+        """bookmark & label findings at all scopes"""
+        symbol_table = currentProgram().getSymbolTable()  # type: ignore [name-defined] # noqa: F821
+        name_space = self.create_namespace()
+
+        for addr in self.locations:
+            txt = self.capability.replace(" ", "-")
+            a = toAddr(hex(addr))  # type: ignore [name-defined] # noqa: F821
+
+            add_bookmark(a, txt)
+
+            # avoid re-naming functions
+            # to namespace/rule names
+            to_cont = False
+            for sym in symbol_table.getSymbols(a):
+                if sym.getSymbolType() == SymbolType.FUNCTION:
+                    to_cont = True
+                    txt = sym.getName()
+                    # label to classify function under capa-generated namespace
+                    createLabel(a, txt, name_space, True, SourceType.USER_DEFINED)  # type: ignore [name-defined] # noqa: F821
+
+            if to_cont:
+                continue
```

Comment:
Let's rename `to_cont` to something more descriptive, this check is confusing as it currently stands.